crates/ruff_linter/src/rules/pylint/rules/binary_op_and_normal_assignment.rs

```diff
@@ -0,0 +1,230 @@
+use ast::{Expr, StmtAugAssign};
+use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for normal assignment statements whose target is the same as the
+/// left operand of a binary operator, in which cases, augmented assignment
+/// could potentially be used instead.
+///
+/// ## Why is this bad?
+/// Augmented assignment operators are more concise to perform a binary
+/// operation and assign results back to one of the operands.
+///
+/// ## Example
+/// ```python
+/// a = a + 1
+/// ```
+///
+/// Use instead:
+/// ```python
+/// a += 1
+/// ```
+///
+/// ## Fix safety
+/// This rule's fix is marked as being unsafe, in that it could alter semantics
+/// of the given Python code in some scenarios.
+///
+/// For example, the following code using mutable data types as the assignment
+/// target
+/// ```python
+/// a = [1]
+/// b = a
+/// a = a + [2]
+/// assert (a, b) == ([1, 2], [1])
+/// ```
+///
+/// is not the same as
+/// ```python
+/// a = [1]
+/// b = a
+/// a += [2]
+/// assert (a, b) == ([1, 2], [1, 2])
+/// ```
+#[violation]
+pub struct BinaryOpAndNormalAssignment;
+
+impl Violation for BinaryOpAndNormalAssignment {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Normal assignment with left operand of binary operator being the same as the target."
+        )
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use augmented assignment instead.".to_string())
+    }
+}
+
+pub(crate) fn binary_op_and_normal_assignment(
+    checker: &mut Checker,
+    assign @ ast::StmtAssign { value, targets, .. }: &ast::StmtAssign,
+) {
+    if targets.len() != 1 {
+        return;
+    }
+    let target = targets.first().unwrap();
+
+    let rhs_expr = value
+        .as_bin_op_expr()
+        .map(|e| (e.left.as_ref(), e.op, e.right.as_ref()));
+    if rhs_expr.is_none() {
+        return;
+    }
+    let (left_operand, operator, right_operand) = rhs_expr.unwrap();
+
+    if name_expr(target, left_operand)
```

Comment:
Instead of testing these conditions, you could match the target/left_operand type with match.

There may be some other rules to serve as inspiration e.g.
https://github.com/astral-sh/ruff/blob/235cfb79769da2c435b9c88d8bae4a79f1234857/crates/ruff_linter/src/rules/flake8_bugbear/rules/static_key_dict_comprehension.rs#L73